pageserver/src/page_service.rs

```diff
@@ -358,26 +374,58 @@ impl PageServerHandler {
             auth,
             claims: None,
             connection_ctx,
+            shard_timelines: HashMap::new(),
         }
     }
 
-    /// Wrap PostgresBackend::flush to respect our CancellationToken: it is important to use
-    /// this rather than naked flush() in order to shut down promptly.  Without this, we would
-    /// block shutdown of a tenant if a postgres client was failing to consume bytes we send
-    /// in the flush.
+    /// Analogous to calling cancelled() on a Timeline's cancellation token: waits for cancellation.
+    ///
+    /// We use many Timeline objects, and hold GateGuards on all of them.  We must therefore respect
+    /// all of their cancellation tokens.
+    async fn timeline_cancelled(&self) {
+        // A short wait before we expend the cycles to walk our timeline map.  This avoids incurring
+        // that cost every time we check for cancellation.
+        tokio::time::sleep(Duration::from_millis(10)).await;
+
+        let mut futs = self
+            .shard_timelines
+            .values()
+            .map(|ht| ht.timeline.cancel.cancelled())
+            .collect::<FuturesUnordered<_>>();
+
+        futs.next().await;
+    }
+
+    /// Analogous to calling is_cancelled() on a Timeline's cancellation token
+    fn timeline_is_cancelled(&self) -> bool {
+        self.shard_timelines
+            .values()
+            .any(|ht| ht.timeline.cancel.is_cancelled() || ht.timeline.is_stopping())
+    }
+
     async fn flush_cancellable<IO>(
         &self,
         pgb: &mut PostgresBackend<IO>,
-        cancel: &CancellationToken,
+        cancel: Option<&CancellationToken>,
     ) -> Result<(), QueryError>
     where
         IO: AsyncRead + AsyncWrite + Send + Sync + Unpin,
     {
+        let general_cancel = async {
+            match cancel {
+                Some(c) => c.cancelled().await,
+                None => task_mgr::shutdown_token().cancelled().await,
+            }
+        };
```

Comment:
Why is this either-or and not "both"?

Reply:
The intent was that `cancel` would be a tenant or timeline cancellation token, which is always subordinate to the overall cancellation of the service, so checking both is unnecessary.

Fortunately this wrinkle goes away in 013a8ab1f where I make cancel non-optional, so that there's no need for a fallback: we just always use an explicit cancellation token.